### PR TITLE
 HLM-311 | Adding a consumer for Registry Async Behaviour

### PR DESCRIPTION
**Pull request by ajay-beehyv**
@prasanna-egov @saiprakash-v @egov-joy

### CODE REVIEW

File: backend/registry-async-handler/main.js [hunk 1]
```diff
@@ -0,0 +1,36 @@
+const {Kafka} = require('kafkajs');
+const axios = require("axios");
+
+async function init_handler(conf){
+    const kafka = new Kafka({
+        clientId: 'async-handler',
```
Comment (prasanna-egov) [inline backend/registry-async-handler/main.js:6]: Registry Async Handler please

File: backend/registry-async-handler/main.js [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+        brokers: conf.KAFKA_BOOTSTRAP_SERVER.split(",")
+    });
+    init_postCreateEntity_consumer(conf,kafka);
+}
+
+async function init_postCreateEntity_consumer(conf,kafka){
```
Comment (prasanna-egov) [inline backend/registry-async-handler/main.js:12]: be consistent either use camelCase or _, don't mix

File: backend/registry-async-handler/main.js [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    const consumer = kafka.consumer({ groupId: 'post_create_entity', sessionTimeout: conf.KAFKA_CONSUMER_SESSION_TIMEOUT });
+    await consumer.connect();
+    
+    await consumer.subscribe({topic: conf.POST_CREATE_ENTITY_TOPIC, fromBeginning: true});
+    const entityType = conf.OSID_ENTITY_TYPE;
+    await consumer.run({
+    eachMessage: async ({ message}) => {
+      const certificateOsidMessage = JSON.parse(message.value.toString());
+      console.log( {entitytype: certificateOsidMessage.entityType});
+      if(certificateOsidMessage.entityType == conf.CERTIFICATE_ENTITY_TYPE){
```
Comment (prasanna-egov) [inline backend/registry-async-handler/main.js:23]: Where is this conf from

File: backend/registry-async-handler/main.js [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+        try{
+          const certificateOsidMapResponse = await  axios.post(`${conf.SUNBIRD_CERTIFICATE_URL}${entityType}`, JSON.parse(message.value.toString()),
```
Comment (prasanna-egov) [inline backend/registry-async-handler/main.js:25]: move to services folder. Follow similar forlder structure as other microservices.

File: backend/registry-async-handler/package.json [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+    "test": "jest"
+  },
+  "repository": {
+    "type": "git",
+    "url": "https://github.com/egovernments/DIVOC/tree/main/backend/registry_async_handler"
+  },
+  "author": "divoc",
+  "license": "GPL-3.0-or-later",
+  "dependencies": {
+    "axios": "^0.21.1",
```
Comment (prasanna-egov) [inline backend/registry-async-handler/package.json:17]: use latest version of axios

**Review by prasanna-egov — CHANGES_REQUESTED** [review]
Docker compose entry is missing. Dockerfile and Makefile are missing. Add those as well.

**Comment by egov-joy** [comment]
@ajay-beehyv please mention the story number in the commit message

**Comment by prasanna-egov** [comment]
There are a lot of changes that are needed. Env variables need to move to config file. Main file need not have a module exports. I am sure this entire thing is not working. Could you do dev testing before raising the PR.

**Comment by ajay-beehyv** [comment]
@prasanna-egov @egov-joy @saiprakash-v 
 This PR containes registry-async-handler microservice, schema for transactionOsidMap, APi to get osid from tranactionId in vc-management-service.
Removed divoc listener from this PR to reduce complexity and added new issue for it (HLM-477)

File: backend/vc-management-service/src/configs/config.js [hunk 6]
```diff
@@ -6,6 +6,7 @@ const JWT_PUBLIC_KEY = process.env.AUTH_PUBLIC_KEY;
 const BASE_URL = process.env.ISSUER_SERVICE_BASE_URL || '/vc-management/';
 const REQUEST_TIMEOUT = process.env.REQUEST_TIMEOUT || 10000;
 const ISSUER_NAME = process.env.ISSUER_NAME || 'Issuer';
+const STORED_ENTITY_TYPE = process.env.STORED_ENTITY_TYPE ;
```
Comment (saiprakash-v) [inline backend/vc-management-service/src/configs/config.js:9]: This is just a table name. Can't we have it as a constant and not config?

File: backend/vc-management-service/src/services/sunbird.service.js [hunk 7]
```diff
@@ -86,10 +86,31 @@ const getIssuerId = async(token) => {
             });
 }
 
+const getTransaction = async (transactionId, token) => {
+    console.log({transctionUrl: constants.SUNBIRD_GET_TRANSACTION_URL, 
+                transactionId: transactionId});
+    const transactionRequest = {
+        id:  "sunbird-rc.registry.search",
```
Comment (prasanna-egov) [inline backend/vc-management-service/src/services/sunbird.service.js:93]: Move to constants

File: backend/vc-management-service/src/services/sunbird.service.js [hunk 8]
```diff
@@ -86,10 +86,31 @@ const getIssuerId = async(token) => {
...
+        ver: "1.0",
+        ets: "",
```
Comment (prasanna-egov) [inline backend/vc-management-service/src/services/sunbird.service.js:95]: What is this ets.

Comment (ajay-beehyv) [reply backend/vc-management-service/src/services/sunbird.service.js:95]: used that approach fron divoc/certificate-api, I'll test if all those variable are necessary

Comment (ajay-beehyv) [reply backend/vc-management-service/src/services/sunbird.service.js]: Removed unnecessary field from Transaction Request

**Comment by ajay-beehyv** [comment]
> Docker compose entry is missing. Dockerfile and Makefile are missing. Add those as well.

resolved

**Pull request merged**